numba/core/untyped_passes.py

```diff
@@ -1426,6 +1427,152 @@ def run_pass(self, state):
+                        continue
+
+                    for arg in value.args:
+                        # check if any of the args to isinstance is a PHI node
+                        iv = func_ir._definitions[arg.name]
+                        assert len(iv) == 1  # SSA!
+                        if isinstance(iv[0], ir.Expr) and iv[0].op == 'phi':
+                            msg = (f'isinstance() cannot determine the '
+                                   f'type of {arg.name} due to a branch')
+                            raise errors.NumbaTypeError(msg, loc=assign.loc)
+
+                lit = typemap.get(target.name, None)
+                if lit and isinstance(lit, types.Literal):
+                    # replace assign instruction by ir.Const(lit) iff
+                    # lit is a literal value
+                    rhs = ir.Const(lit.literal_value, assign.loc)
+                    new_assign = ir.Assign(rhs, target, assign.loc)
+
+                    # replace instruction
+                    block.insert_after(new_assign, assign)
+                    block.remove(assign)
+
+                    changed = True
+
+        # reset type inference now we are done with the partial results
+        state.typemap = None
+        state.calltypes = None
+
+        if changed:
+            # Rebuild definitions
+            func_ir._definitions = build_definitions(func_ir.blocks)
+
+        return changed
+
+
+@register_pass(mutates_CFG=True, analysis_only=False)
+class LiteralPropagationSubPipelinePass(FunctionPass):
+    """Implement literal propagation based on partial type inference"""
+    _name = "LiteralPropagation"
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+
+    def run_pass(self, state):
+        # Determine whether to even attempt this pass... if there's no
+        # `isinstance` as a global or as a freevar then just skip.
+        found = False
+        func_ir = state.func_ir
+        for blk in func_ir.blocks.values():
+            for asgn in blk.find_insts(ir.Assign):
+                if isinstance(asgn.value, (ir.Global, ir.FreeVar)):
+                    if asgn.value.value is isinstance:
+                        found = True
+                        break
+            if found:
+                break
+        if not found:
+            return False
+
+        # run as subpipeline
+        from numba.core.compiler_machinery import PassManager
+        from numba.core.typed_passes import PartialTypeInference
+        pm = PassManager("literal_propagation_subpipeline")
+
+        pm.add_pass(PartialTypeInference, "performs partial type inference")
+        pm.add_pass(PropagateLiterals, "performs propagation of literal values")
+
+        # rewrite consts / dead branch pruning
+        pm.add_pass(RewriteSemanticConstants, "rewrite semantic constants")
+        pm.add_pass(DeadBranchPrune, "dead branch pruning")
+        pm.add_pass(DeadCodeElimination, "dead code elimination")
```

Comment:
What's the reason for adding this? I've a vague memory of it being unsafe for use on IR that isn't fully typed as alias analysis is needed.

Reply:
Would it be ok to add support for namedtuples in a follow up PR?